File: backend/wasm/mopsa_worker.js
/**
 * mopsa_worker.js
 *
 * Web Worker that executes the Mopsa WASM analysis in a background thread
 * so the main UI thread stays responsive during analysis.
 *
 * The WASM binary (15 MB) and data file (12.6 MB) are fetched ONCE when the
 * worker starts. Each run re-instantiates the cached WebAssembly.Module
 * (fast, no network) because the OCaml runtime cannot be re-entered.
 *
 * Two execution modes
 * ───────────────────
 * 1. Batch (-engine=automatic) — request/response, output buffered and
 *    returned at the end. This is the original one-shot model.
 *      Receives: { type:'analyze', id, options, code, config, codeFile, extraFiles }
 *      Posts:    { type:'result',  id, output }
 *
 * 2. Session (-engine=interactive | dap) — ONE long-lived run that blocks on
 *    synchronous stdin. stdin is fed over a SharedArrayBuffer channel
 *    (sync-message.js) via Atomics.wait, because the worker thread is frozen
 *    inside the blocking read and cannot service postMessage. stdout/stderr are
 *    streamed back as raw bytes as they are produced (the worker is never
 *    blocked on output).
 *      Receives: { type:'start', id, engine, options, code, config, codeFile,
 *                  extraFiles, stdinChannel }
 *      Posts:    { type:'started',      id }
 *                { type:'stdout-bytes', id, bytes }      // raw, transferable
 *                { type:'session-ended', id, code }
 *                { type:'session-error', id, message }
 *
 * A session cannot be interrupted via postMessage (the thread is blocked in
 * Atomics.wait); the main thread kills it with worker.terminate() + respawn.
 */
"use strict";

importScripts("./ocamlrun.js", "./sync-message.js");

var _wasmModulePromise = (
  typeof WebAssembly.compileStreaming === "function"
    ? WebAssembly.compileStreaming(fetch("./ocamlrun.wasm"))
    : fetch("./ocamlrun.wasm")
        .then(function (r) {
          return r.arrayBuffer();
        })
        .then(function (bytes) {
          return WebAssembly.compile(bytes);
        })
).catch(function (e) {
  console.error("[Mopsa Worker] WASM pre-compilation failed:", e);
  return null;
});

var _dataBufferPromise = fetch("./ocamlrun.data")
  .then(function (r) {
    return r.arrayBuffer();
  })
  .catch(function (e) {
    console.error("[Mopsa Worker] Data file pre-fetch failed:", e);
    return null;
  });

// ── Shared helpers ──────────────────────────────────────────────────────────

function buildArgs(options, isHelp) {
  return ["build/mopsa.bc"]
    .concat(isHelp ? [] : ["-config", "/config.json"])
    .concat(["-share-dir", "/share/mopsa", "-I", "/clang-headers", "-I", "/usr/include"])
    .concat(options || []);
}

// Write code/config/extra files into the virtual FS, and set TERM so Mopsa
// emits ANSI colors (utils/core/debug.ml reads Sys.getenv "TERM"; the Worker
// has none by default → colorless output). Requires 'ENV' exported (Makefile).
function makePreRun(code, config, codeFile, extraFiles) {
  return function (M) {
    if (M.ENV) M.ENV.TERM = "xterm-256color";

    M.FS.writeFile("/config.json", config);

    var dir = codeFile.substring(0, codeFile.lastIndexOf("/"));
    if (dir && dir !== "/") {
      try {
        M.FS.mkdirTree(dir);
      } catch (_) {}
    }

    Object.keys(extraFiles || {}).forEach(function (path) {
      var d = path.substring(0, path.lastIndexOf("/"));
      if (d && d !== "/") {
        try {
          M.FS.mkdirTree(d);
        } catch (_) {}
      }
      M.FS.writeFile(path, extraFiles[path]);
    });

    // Write code file LAST so it always overrides any stale extraFiles entry.
    M.FS.writeFile(codeFile, code);
  };
}

function withWasm(moduleConfig, wasmModule, dataBuffer, onInstantiateError) {
  if (wasmModule) {
    moduleConfig.instantiateWasm = function (imports, successCallback) {
      WebAssembly.instantiate(wasmModule, imports)
        .then(function (instance) {
          successCallback(instance, wasmModule);
        })
        .catch(onInstantiateError);
      return {};
    };
  }
  if (dataBuffer) {
    moduleConfig.getPreloadedPackage = function () {
      return dataBuffer;
    };
  }
  return moduleConfig;
}

// ── Batch mode (-engine=automatic) ────────────────────────────────────────────

function runBatch(msg) {
  var id = msg.id;
  var options = msg.options || [];
  var output = "";
  var isHelp = options.indexOf("-help") !== -1;

  Promise.all([_wasmModulePromise, _dataBufferPromise]).then(function (results) {
    var moduleConfig = withWasm(
      {
        arguments: buildArgs(options, isHelp),
        print: function (line) {
          output += line + "\n";
        },
        printErr: function (line) {
          output += line + "\n";
        },
        preRun: [makePreRun(msg.code, msg.config, msg.codeFile, msg.extraFiles)],
      },
      results[0],
      results[1],
      function (e) {
        self.postMessage({
          type: "result",
          id: id,
          output: output + "\n[WASM instantiation error] " + ((e && e.message) || e),
        });
      },
    );

    createMopsaModule(moduleConfig)
      .then(function () {
        self.postMessage({ type: "result", id: id, output: output });
      })
      .catch(function (e) {
        if (e && typeof e === "object" && "status" in e) {
          // Normal OCaml exit() output is already captured.
          self.postMessage({ type: "result", id: id, output: output });
        } else {
          self.postMessage({
            type: "result",
            id: id,
            output: output + "\n[WASM error] " + ((e && e.message) || e),
          });
        }
      });
  });
}

// ── Session mode (-engine=interactive | dap) ──────────────────────────────────

// Synchronous stdin: blocks the worker thread on Atomics.wait until the main
// thread writes a message over the SAB channel. Returns one byte per call,
// or null. Emscripten's createDevice read() loops calling this until it gets
// null, so we must return null at the END of each chunk to let read() return
// (otherwise it would block waiting for a second line the user hasn't typed).
// The `delivering` flag distinguishes "just finished a chunk" (return null,
// don't block) from "start of a fresh read" (block for the next message).
//
// `flushOut` is called right before we block: while the thread sits in
// Atomics.wait, microtasks do NOT run, so any buffered stdout (e.g. the
// no-newline "mopsa >> " prompt printed just before this read) would never
// reach the main thread and the user would face a silent prompt. Flushing
// synchronously here guarantees the prompt/response is shown before we wait.
function makeStdin(channel, flushOut) {
  var buf = new Uint8Array(0);
  var pos = 0;
  var eof = false;
  var delivering = false;
  var msgId = 0;
  var encoder = new TextEncoder();
  return function () {
    if (pos < buf.length) {
      delivering = true;
      return buf[pos++];
    }
    if (eof) return null; // permanent EOF
    if (delivering) {
      delivering = false; // end of the current chunk → let read() return
      return null;
    }
    // Start of a new read: flush pending output, then block for the next chunk.
    flushOut();
    while (true) {
      var m = self.syncMessage.readMessage(channel, String(msgId++), {});
      if (m == null) continue; // defensive: only on timeout/interrupt
      if (m.eof) {
        eof = true;
        return null;
      }
      buf = encoder.encode(m.data || "");
      pos = 0;
      if (buf.length === 0) continue; // empty payload → wait for real data
      delivering = true;
      return buf[pos++];
    }
  };
}

// Raw stdout/stderr: collect bytes and post them. Batches within a synchronous
// run via a microtask, but also exposes a synchronous flush() that makeStdin
// calls before blocking. Used for BOTH engines: interactive feeds the bytes to
// xterm (no-newline prompts must stream immediately); dap reassembles
// Content-Length frames from them.
function makeByteSink(id) {
  var buf = [];
  var scheduled = false;
  function flush() {
    scheduled = false;
    if (!buf.length) return;
    var bytes = new Uint8Array(buf);
    buf = [];
    self.postMessage({ type: "stdout-bytes", id: id, bytes: bytes }, [bytes.buffer]);
  }
  var sink = function (val) {
    if (val === null) {
      flush();
      return;
    }
    buf.push(val & 0xff);
    if (!scheduled) {
      scheduled = true;
      queueMicrotask(flush);
    }
  };
  sink.flush = flush;
  return sink;
}

function runSession(msg) {
  var id = msg.id;
  var engine = msg.engine; // 'interactive' | 'dap'
  var channel = msg.stdinChannel;
  // Worker is authoritative for the engine flag; callers must NOT also pass it.
  var options = (msg.options || []).concat(["-engine=" + engine]);

  if (!channel) {
    self.postMessage({
      type: "session-error",
      id: id,
      message: "No stdin channel (SharedArrayBuffer unavailable?)",
    });
    return;
  }

  var sink = makeByteSink(id);

  Promise.all([_wasmModulePromise, _dataBufferPromise]).then(function (results) {
    var moduleConfig = withWasm(
      {
        arguments: buildArgs(options, false),
        stdin: makeStdin(channel, sink.flush),
        stdout: sink,
        stderr: sink,
        preRun: [makePreRun(msg.code, msg.config, msg.codeFile, msg.extraFiles)],
      },
      results[0],
      results[1],
      function (e) {
        self.postMessage({
          type: "session-error",
          id: id,
          message: "[WASM instantiation error] " + ((e && e.message) || e),
        });
      },
    );

    self.postMessage({ type: "started", id: id });

    createMopsaModule(moduleConfig)
      .then(function () {
        self.postMessage({ type: "session-ended", id: id, code: 0 });
      })
      .catch(function (e) {
        if (e && typeof e === "object" && "status" in e) {
          self.postMessage({ type: "session-ended", id: id, code: e.status });
        } else {
          self.postMessage({
            type: "session-error",
            id: id,
            message: "[WASM error] " + ((e && e.message) || e),
          });
        }
      });
  });
}

// ── Dispatch ──────────────────────────────────────────────────────────────────

self.onmessage = function (event) {
  var msg = event.data;
  if (!msg) return;
  if (msg.type === "analyze") runBatch(msg);
  else if (msg.type === "start") runSession(msg);
  // 'kill' is unreachable while the thread is blocked in Atomics.wait; the main
  // thread kills a session with worker.terminate() + respawn instead.
};

console.log("[Mopsa Worker] ready");

File: backend/jsoo/mopsa_worker.ml
(* mopsa_worker.ml — js_of_ocaml backend worker.
 *
 * Web Worker that runs the Mopsa analysis compiled to pure JavaScript
 * (js_of_ocaml), as a lighter alternative to the WASM backend. Feature
 * differences vs WASM: no C / C+Python analysis (the Clang parser is a
 * C++ library) and no Apron-based relational domains (C library; their
 * primitives raise if a config selects them).
 *
 * The postMessage protocol is the same as backend/wasm/mopsa_worker.js:
 *
 * 1. Batch:
 *      Receives: { type:'analyze', id, options, code, config, codeFile, extraFiles }
 *      Posts:    { type:'result',  id, output }
 *
 * 2. Session (interactive | dap): one long-lived run, stdin fed
 *    synchronously over a SharedArrayBuffer channel (sync-message.js),
 *    stdout/stderr streamed back as raw bytes.
 *      Receives: { type:'start', id, engine, options, code, config, codeFile,
 *                  extraFiles, stdinChannel }
 *      Posts:    { type:'started', id } { type:'stdout-bytes', id, bytes }
 *                { type:'session-ended', id, code } { type:'session-error', id, message }
 *
 * The OCaml runtime state persists across runs, so the API shim
 * (mopsa_api_jsoo.js) respawns the worker after every run to get a
 * fresh analyzer state.
 *
 * The share directory (configs + stubs) is NOT baked into the bundle:
 * the API shim fetches the frontend's share.json (already generated for
 * the presets UI) and passes the relevant files in each message as
 * `shareFiles` (path relative to the share dir → content); they are
 * written under /static/share/mopsa before the run.
 *)

open Js_of_ocaml

let share_dir = "/static/share/mopsa"

(* ── Virtual filesystem helpers ─────────────────────────────────────────── *)

let write_file ~name ~content =
  let open Sys_js in
  try
    let _ = read_file ~name in
    update_file ~name ~content
  with Sys_error _ -> create_file ~name ~content

let setup_files ~code ~config ~code_file ~extra_files ~share_files =
  List.iter
    (fun (path, content) ->
      write_file ~name:(share_dir ^ "/" ^ path) ~content)
    share_files;
  write_file ~name:"/config.json" ~content:config;
  List.iter (fun (path, content) -> write_file ~name:path ~content) extra_files;
  (* code file last so it overrides any stale extraFiles entry *)
  write_file ~name:code_file ~content:code

(* ── Message field access ───────────────────────────────────────────────── *)

let get_str msg field : string = Js.to_string (Js.Unsafe.get msg field)

let get_int msg field : int = int_of_float (Js.float_of_number (Js.Unsafe.get msg field))

let get_options msg : string list =
  let arr : Js.js_string Js.t Js.js_array Js.t = Js.Unsafe.get msg "options" in
  Js.to_array arr |> Array.to_list |> List.map Js.to_string

(* {path → content} JS object field → assoc list *)
let get_file_map msg field : (string * string) list =
  let obj = Js.Unsafe.get msg field in
  if Js.to_string (Js.typeof obj) = "object" then
    Js.object_keys obj |> Js.to_array |> Array.to_list
    |> List.map (fun k -> (Js.to_string k, Js.to_string (Js.Unsafe.get obj k)))
  else []

let post (fields : (string * Js.Unsafe.any) list) =
  Worker.post_message (Js.Unsafe.obj (Array.of_list fields))

let str s = Js.Unsafe.inject (Js.string s)
let num i = Js.Unsafe.inject (Js.number_of_float (float_of_int i))

(* ── Running the analyzer ───────────────────────────────────────────────── *)

let build_argv options =
  let is_help = List.mem "-help" options in
  let base = if is_help then [ "mopsa" ] else [ "mopsa"; "-config"; "/config.json" ] in
  Array.of_list (base @ [ "-share-dir"; share_dir ] @ options)

let run_analysis options : int =
  Mopsa_analyzer.Framework.Runner.parse_options (build_argv options)
    Mopsa_analyzer.Framework.Runner.analyze_files ()

(* ── Batch mode ─────────────────────────────────────────────────────────── *)

let run_batch msg =
  let id = get_int msg "id" in
  let buf = Buffer.create 4096 in
  Sys_js.set_channel_flusher stdout (Buffer.add_string buf);
  Sys_js.set_channel_flusher stderr (Buffer.add_string buf);
  setup_files ~code:(get_str msg "code") ~config:(get_str msg "config")
    ~code_file:(get_str msg "codeFile")
    ~extra_files:(get_file_map msg "extraFiles")
    ~share_files:(get_file_map msg "shareFiles");
  (try
     let (_ : int) = run_analysis (get_options msg) in
     ()
   with e ->
     Buffer.add_string buf
       (Printf.sprintf "\n[jsoo error] %s\n" (Printexc.to_string e)));
  (try flush Stdlib.stdout; flush Stdlib.stderr with _ -> ());
  post [ ("type", str "result"); ("id", num id); ("output", str (Buffer.contents buf)) ]

(* ── Session mode (interactive | dap) ───────────────────────────────────── *)

(* Convert an OCaml string (raw bytes) to a Uint8Array without any
   UTF-8 re-encoding, so DAP Content-Length framing stays byte-exact. *)
let bytes_of_string (s : string) : Js.Unsafe.any =
  let n = String.length s in
  let arr = Js.Unsafe.new_obj Js.Unsafe.global##.Uint8Array [| num n |] in
  for i = 0 to n - 1 do
    Js.Unsafe.set arr i (Char.code s.[i])
  done;
  Js.Unsafe.inject arr

let run_session msg =
  let id = get_int msg "id" in
  let engine = get_str msg "engine" in
  let channel = Js.Unsafe.get msg "stdinChannel" in
  let options = get_options msg @ [ "-engine=" ^ engine ] in

  (try Worker.import_scripts [ "./sync-message.js" ]
   with _ ->
     post
       [ ("type", str "session-error"); ("id", num id);
         ("message", str "cannot load sync-message.js") ]);

  let post_bytes s =
    if String.length s > 0 then
      post [ ("type", str "stdout-bytes"); ("id", num id); ("bytes", bytes_of_string s) ]
  in
  Sys_js.set_channel_flusher stdout post_bytes;
  Sys_js.set_channel_flusher stderr post_bytes;

  (* Synchronous stdin: flush any pending prompt, then block on the
     SharedArrayBuffer channel until the main thread writes a message
     ({data: string} or {eof: true}, cf. mopsa_api sendInput/sendEof). *)
  let msg_id = ref 0 in
  Sys_js.set_channel_filler stdin (fun () ->
      (try flush Stdlib.stdout; flush Stdlib.stderr with _ -> ());
      let m =
        Js.Unsafe.fun_call
          (Js.Unsafe.get (Js.Unsafe.global##.syncMessage) "readMessage")
          [| channel;
             Js.Unsafe.inject (Js.string (string_of_int !msg_id));
             Js.Unsafe.obj [||] |]
      in
      incr msg_id;
      let eof = Js.Optdef.test (Js.Unsafe.get m "eof")
                && Js.to_bool (Js.Unsafe.get m "eof") in
      if eof then "" (* empty string = EOF *)
      else
        let data = Js.Unsafe.get m "data" in
        if Js.Optdef.test data then Js.to_string data else "");

  setup_files ~code:(get_str msg "code") ~config:(get_str msg "config")
    ~code_file:(get_str msg "codeFile")
    ~extra_files:(get_file_map msg "extraFiles")
    ~share_files:(get_file_map msg "shareFiles");

  post [ ("type", str "started"); ("id", num id) ];
  (try
     let code = run_analysis options in
     (try flush Stdlib.stdout; flush Stdlib.stderr with _ -> ());
     post [ ("type", str "session-ended"); ("id", num id); ("code", num code) ]
   with e ->
     (try flush Stdlib.stdout; flush Stdlib.stderr with _ -> ());
     post
       [ ("type", str "session-error"); ("id", num id);
         ("message", str (Printexc.to_string e)) ])

(* ── Dispatch ───────────────────────────────────────────────────────────── *)

let () =
  Printexc.record_backtrace true;
  Worker.set_onmessage (fun msg ->
      match get_str msg "type" with
      | "analyze" -> run_batch msg
      | "start" -> run_session msg
      | _ -> ());
  Console.console##log (Js.string "[Mopsa jsoo Worker] ready")
